### ISSUE TITLE
Gradient Field CNN: Export artifacts + push to Hugging Face Hub (fusion-ready)

### ISSUE THREAD

**Issue by lukhsaankumar** [McMasterAI-Society/DeepFakeDetector#45]
## Issue Type

- [ ] **Model**: ML model bug, training issue, or architecture problem
- [ ] **Data**: Dataset issue, preprocessing bug, or data pipeline problem  
- [ ] **Web**: Frontend bug or UI issue in the Next.js dashboard
- [ ] **API**: Backend API bug or FastAPI endpoint issue
- [ ] **Research**: Research question or experimental feature request
- [ ] **Documentation**: Documentation bug or improvement needed
- [ ] **Bug**: General bug fix needed
- [x] **Enhancement**: New feature or improvement request

## Description
Standardize the **Gradient Field CNN** export workflow and publish all required artifacts to Hugging Face Hub so the fusion model + backend can load it remotely and deterministically.

This ticket updates the workflow from:
- **Previous**: `save artifacts locally under models/grad_field_cnn/`
- **New**: `export artifacts → upload to Hugging Face → backend downloads + loads for inference`

**Hugging Face repo**: DeepFakeDetector/gradfield-cnn (https://huggingface.co/DeepFakeDetector/gradfield-cnn)

## Acceptance Criteria
- [ ] A Hugging Face **Write token** is created and available locally as an environment variable (`HF_TOKEN_WRITE`)
- [ ] `gradient_field_cnn_baseline.ipynb` contains a final **Export + Upload** section
- [ ] Artifacts are exported with the standardized layout:
  - `model.pth` (PyTorch state_dict)
  - `config.json` (architecture + threshold + gradient-field settings)
  - `preprocess.json` (input size, normalization, channels, etc.)
  - `label_map.json` (0 = real, 1 = fake)
  - `README.md` (model card)
- [ ] Gradient-field feature generation settings are explicitly documented in `config.json`
- [ ] All artifacts are uploaded to the Hugging Face repo
- [ ] Notebook includes a minimal reload test (download → load → eval)

### Hugging Face (canonical storage)
Uploaded to: https://huggingface.co/DeepFakeDetector/gradfield-cnn

Must contain:
- `model.pth`
- `config.json`
- `preprocess.json`
- `label_map.json`
- `README.md`

### Notebook update
`gradient_field_cnn_baseline.ipynb` includes cells that:
1. Export all artifacts (including gradient-field params)
2. Upload them to Hugging Face

### Label Convention (CRITICAL)
**Label convention is fixed and shared across all models**:
- `0 = real`
- `1 = fake`

### Gradient Field Requirements
Gradient Field CNN requires a reproducible definition of:
- How gradients are computed (Sobel/Scharr/custom)
- Whether magnitude/direction is used
- Normalization/scaling strategy
- Whether gradients are concatenated with RGB or used alone
- Final channel layout (e.g., 1ch/2ch/3ch/6ch)

### Output
- Fusion will consume `prob_fake` from this model

---

### 1) Create Hugging Face Write Token
Create a write token in Hugging Face account settings and export locally:

**Environment variable (recommended):**
```bash
export HF_TOKEN_WRITE=hf_xxxxxxxxxxxxxxxxx
```

**Never commit this token to git.**

### 2) Notebook Cell Snippets to Add
Add these cells to the end of `gradient_field_cnn_baseline.ipynb`:

#### Cell A — Define gradient-field settings (single source of truth)
Edit these to exactly match your current implementation in the notebook.
```python
GRADFIELD_SETTINGS = {
    # how gradients are generated:
    "operator": "sobel",             # "sobel" | "scharr" | "prewitt" | "custom"
    "kernel_size": 3,
    "use_magnitude": True,           # True => sqrt(gx^2 + gy^2)
    "use_direction": False,          # True => atan2(gy, gx)
    "use_gx_gy_channels": True,      # True => keep gx and gy separately

    # how gradients are combined with image:
    "input_mode": "grad_only",       # "grad_only" | "rgb_plus_grad"
    "rgb_plus_grad_concat_order": "rgb_first",  # if used

    # scaling / normalization:
    "grad_scale": "minmax_0_1",      # "minmax_0_1" | "standardize" | "none"
    "epsilon": 1e-6,

    # channel expectations:
    "expected_channels": 2           # e.g., gx, gy (2) or magnitude-only (1) or rgb+gx+gy (5)
}
```

#### Cell B — Export artifacts (weights + configs)
```python
import os, json, torch

EXPORT_DIR = "export/gradfield-cnn"
os.makedirs(EXPORT_DIR, exist_ok=True)

# Save weights (state_dict)
torch.save(model.state_dict(), os.path.join(EXPORT_DIR, "model.pth"))

# Preprocessing details (MUST match training)
# NOTE: normalize here refers to what you apply to the model input AFTER gradient-field generation.
preprocess = {
    "input_size": 224,
    "center_crop": True,
    "normalize": {
        "mean": [0.5] * GRADFIELD_SETTINGS["expected_channels"],
        "std":  [0.5] * GRADFIELD_SETTINGS["expected_channels"]
    }
}

label_map = {"0": "real", "1": "fake"}

config = {
    "name": "gradfield-cnn",
    "framework": "pytorch",
    "arch": "resnet18",          # change to your actual backbone
    "num_classes": 2,
    "threshold": 0.50,
    "labels": label_map,

    # CRITICAL: gradient-field generation parameters
    "gradfield_settings": GRADFIELD_SETTINGS,

    "notes": "Gradient Field CNN baseline for deepfake detection"
}

with open(os.path.join(EXPORT_DIR, "preprocess.json"), "w") as f:
    json.dump(preprocess, f, indent=2)

with open(os.path.join(EXPORT_DIR, "label_map.json"), "w") as f:
    json.dump(label_map, f, indent=2)

with open(os.path.join(EXPORT_DIR, "config.json"), "w") as f:
    json.dump(config, f, indent=2)

readme = """---
license: apache-2.0
tags:
  - deepfake-detection
  - image-classification
  - cnn
  - gradients
---

# Gradient Field CNN – DeepFakeDetector

CNN trained on gradient-field representations (e.g., Sobel gx/gy, magnitude) for binary deepfake detection.

## Labels
- 0 = real
- 1 = fake

## Gradient Field Settings
This repo includes `gradfield_settings` inside `config.json` which defines:
- gradient operator (sobel/scharr)
- channel construction (gx/gy vs magnitude)
- scaling/normalization
- whether RGB is concatenated
"""

with open(os.path.join(EXPORT_DIR, "README.md"), "w") as f:
    f.write(readme)

print("Exported Gradient Field CNN artifacts to:", EXPORT_DIR)
print("gradfield_settings:", GRADFIELD_SETTINGS)
```

#### Cell C — Upload artifacts to Hugging Face Hub
```python
import os
from huggingface_hub import HfApi

REPO_ID = "DeepFakeDetector/gradfield-cnn"
HF_TOKEN = os.environ.get("HF_TOKEN_WRITE")
assert HF_TOKEN, "HF_TOKEN_WRITE env var not set."

api = HfApi(token=HF_TOKEN)

# Create repo if it doesn't exist
api.create_repo(repo_id=REPO_ID, repo_type="model", exist_ok=True)

api.upload_folder(
    folder_path=EXPORT_DIR,
    repo_id=REPO_ID,
    repo_type="model",
    commit_message="Upload Gradient Field CNN artifacts (weights + configs + gradfield settings)"
)

print(f"Uploaded to https://huggingface.co/{REPO_ID}")
```

#### Cell D — (Optional but recommended) Sanity reload test (download + load)
```python
import torch
from huggingface_hub import snapshot_download

local_repo = snapshot_download(repo_id=REPO_ID, repo_type="model")

state = torch.load(os.path.join(local_repo, "model.pth"), map_location="cpu")
model.load_state_dict(state)
model.eval()

print("Gradient Field CNN weights reloaded successfully from HF")
print("Local snapshot path:", local_repo)
```

## Definition of Done
- [ ] Hugging Face repo contains all required artifacts (model + config + preprocess + labels + README)
- [ ] Notebook can export + upload in one run
- [ ] Reload test passes
- [ ] `config.json` includes `gradfield_settings` that exactly match training-time feature generation
- [ ] Model is ready for fusion + backend consumption
